[AMDGPU][SDAG] Support source modifiers on select integer operands

chrisjbris opened the pull request:
Extend the DAGCombine() for select to directly support fneg and fabs for i32, v2i32 and i64.

This combine enables select operands that are bitmask sign-bit altering operations to be replaced with fneg, fabs or fneg+fabs. The AMDGPU backend supports source modifiers for these fp logic operations and enabling this combine enables more efficient lowering of these kinds of select instructions.

Pre-cursor PR to simplify https://github.com/llvm/llvm-project/pull/140694.

chrisjbris committed:
Add new test for source modifiers on select

chrisjbris committed:
Populate check-lines before patching

chrisjbris committed:
[AMDGPU][SDAG] Support source modifiers as integer on select

Extend the DAGCombine() for select to directly support fneg and fabs
for i32, v2i32 and i64.

chrisjbris committed:
Simplify switch in BitwiseToSrcModifierOp()

chrisjbris committed:
[NFC] Correct typo in BitwiseToSrcModifierOp()

JanekvO commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4830,6 +4830,65 @@ AMDGPUTargetLowering::foldFreeOpFromSelect(TargetLowering::DAGCombinerInfo &DCI,
+    return SDValue();
+
+  SelectionDAG &DAG = DCI.DAG;
+  SDValue LHS = N.getNode()->getOperand(0);
+  SDValue RHS = N.getNode()->getOperand(1);
```
```suggestion
  SDValue LHS = N->getOperand(0);
  SDValue RHS = N->getOperand(1);
```

JanekvO commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4830,6 +4830,65 @@ AMDGPUTargetLowering::foldFreeOpFromSelect(TargetLowering::DAGCombinerInfo &DCI,
+
+  uint64_t Mask = 0;
+  if (VT.isVector()) {
+    SDValue Splat = DAG.getSplatValue(RHS);
```
We don't know whether it's a splat vector or vector with different operands at this point, right?
Furthermore, why do we need this? At this point in the function we already know that `CRHS` has a value  (through the conditional above) and thus know the constant value through `ConstantSDNode`, so is this dead code?

chrisjbris committed:
Fix bitcast type in performSelectCombine()

JanekvO commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4830,6 +4830,65 @@ AMDGPUTargetLowering::foldFreeOpFromSelect(TargetLowering::DAGCombinerInfo &DCI,
+    Mask = CRHS->getZExtValue();
+
+  EVT FVT = IntToFloatVT(VT);
+  SDValue BC = DAG.getNode(ISD::BITCAST, SDLoc(N), FVT, LHS);
```
nit: can `SDLoc SL(N);` and reuse SL

chrisjbris committed:
Respond to first review comments

shiltian commented on llvm/test/CodeGen/AMDGPU/integer-select-source-modifiers.ll:
```diff
@@ -0,0 +1,222 @@
+  ret i64 %select
+}
+;; NOTE: These prefixes are unused and the list is autogenerated. Do not add tests below this line:
+; GFX11-FAKE16: {{.*}}
+; GFX11-TRUE16: {{.*}}
+; GFX7: {{.*}}
+; GFX9: {{.*}}
```
These are empty so you might want to remove them from the prefix list.

chrisjbris committed:
Respond to secon review comments - rename function and correct test

chrisjbris committed:
[NFC] Remove incomplete dag-style comment

LU-JOHN commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4830,6 +4830,59 @@ AMDGPUTargetLowering::foldFreeOpFromSelect(TargetLowering::DAGCombinerInfo &DCI,
+    break;
+  case ISD::OR:
+    if (Mask == 0x80000000u || Mask == 0x8000000000000000u) {
+      SDValue Neg = DAG.getNode(ISD::FNEG, SDLoc(N), FVT, BC);
+      return DAG.getNode(ISD::FABS, SL, FVT, Neg);
```
OR-ing will set the sign-bit and produce a negative value.  FNEG should be the outer operation.

chrisjbris replied:
Corrected, thanks.

LU-JOHN replied:
Thanks for the fix.  Do you know why the OR test results do not change after this fix?  It seems strange to me that the compiler would generate the same output results with or without this fix.  The fix changes the IR's meaning.

chrisjbris committed:
Make test for bitwise src mods more stringent and correct fneg-fabs order

chrisjbris committed:
Reviewer corrections

arsenm commented on llvm/test/CodeGen/AMDGPU/integer-select-source-modifiers.ll:
```diff
@@ -0,0 +1,372 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx700 < %s | FileCheck -check-prefixes=GFX7 %s
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 < %s | FileCheck -check-prefixes=GFX9 %s
+; RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx1100 < %s | FileCheck -check-prefixes=GFX11 %s
```
Add a common check prefix, the simple cases have identical output

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4830,6 +4830,62 @@ AMDGPUTargetLowering::foldFreeOpFromSelect(TargetLowering::DAGCombinerInfo &DCI,
+                                         TargetLowering::DAGCombinerInfo &DCI) {
+
+  unsigned Opc = N.getNode()->getOpcode();
+  if (Opc != ISD::AND && Opc != ISD::XOR && Opc != ISD::AND)
```
Check for AND repeated twice, one should be OR?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4830,6 +4830,62 @@ AMDGPUTargetLowering::foldFreeOpFromSelect(TargetLowering::DAGCombinerInfo &DCI,
+
+  switch (Opc) {
+  case ISD::XOR:
+    if ((Mask == 0x80000000u && VT.getFixedSizeInBits() == 32) ||
+        (Mask == 0x8000000000000000u && VT.getFixedSizeInBits() == 64))
```
Don't hardcode the special values, you can use CRHS->isSignMask() and CRHS->isMaxSignedValue()

arsenm commented on llvm/test/CodeGen/AMDGPU/integer-select-source-modifiers.ll:
```diff
@@ -0,0 +1,372 @@
+; GFX9-NEXT:    v_or_b32_e32 v2, 0x80000000, v2
+; GFX9-NEXT:    v_cmp_eq_u32_e32 vcc, 0, v0
+; GFX9-NEXT:    v_or_b32_e32 v3, 0x80000000, v3
+; GFX9-NEXT:    v_cndmask_b32_e32 v0, v4, v2, vcc
```
The fneg+fabs cases didn't fold, but you missed those anyway

arsenm commented on llvm/test/CodeGen/AMDGPU/integer-select-source-modifiers.ll:
```diff
@@ -0,0 +1,372 @@
+; GFX7:       ; %bb.0:
+; GFX7-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
+; GFX7-NEXT:    v_cmp_eq_u64_e32 vcc, 0, v[0:1]
+; GFX7-NEXT:    v_xor_b32_e32 v3, 0x80000000, v3
+; GFX7-NEXT:    v_cndmask_b32_e32 v0, v4, v2, vcc
+; GFX7-NEXT:    v_cndmask_b32_e32 v1, v5, v3, vcc
```
This didn't fold into the source modifier

arsenm replied:
For this you should break the select somewhere. The sign bits only matter for the top half of the value

chrisjbris replied:
ok, thanks.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4830,6 +4830,62 @@ AMDGPUTargetLowering::foldFreeOpFromSelect(TargetLowering::DAGCombinerInfo &DCI,
+    if ((Mask == 0x7fffffffu && VT.getFixedSizeInBits() == 32) ||
+        (Mask == 0x7fffffffffffffffu && VT.getFixedSizeInBits() == 64))
+      return DAG.getNode(ISD::FABS, SL, FVT, BC);
+    break;
```
The 64-bit case is really a 32-bit operation on the high bits, with passthrough of the low bits

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4830,6 +4830,62 @@ AMDGPUTargetLowering::foldFreeOpFromSelect(TargetLowering::DAGCombinerInfo &DCI,
 
+static EVT getFloatVT(EVT VT) {
+  return VT.isVector() ? MVT::getVectorVT(
+                             MVT::getFloatingPointVT(VT.getScalarSizeInBits()),
+                             VT.getVectorNumElements())
```
changeVectorElementType

arsenm reviewed (commented):
There are 2 ways to go about this. Option 1 is what you've done here, where you hack up the integer operations to use the FP types. Option 2 would be to start directly selecting the modifiers through the integer patterns. There are plusses and minuses to each, and the main deciding factor would be how this interferes with other combines.

The examples here don't have any other context to make a judgement on. We also know have SimplifyDemandedBits through the FP sign bit operators, so maybe it matters less now.

arsenm commented on llvm/test/CodeGen/AMDGPU/integer-select-source-modifiers.ll:
```diff
@@ -0,0 +1,372 @@
+; GFX11-NEXT:    s_setpc_b64 s[30:31]
+  %neg.a = xor i32 %a, u0x80000000
+  %cmp = icmp eq i32 %cond, zeroinitializer
+  %select = select i1 %cmp, i32 %neg.a, i32 %b
```
Should also test with the modifiers in both positions (i.e. commuted versions of all the tests). Also cases where you can fold different modifiers into each input

chrisjbris replied:
I've added some new test cases.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4830,6 +4830,62 @@ AMDGPUTargetLowering::foldFreeOpFromSelect(TargetLowering::DAGCombinerInfo &DCI,
+
+  EVT VT = RHS.getValueType();
+
+  assert((VT == MVT::i32 || VT == MVT::v2i32 || VT == MVT::i64) &&
+         "Expected i32, v2i32 or i64 value type.");
```
In the future could be even fancier and handle v2f16 with an fneg of only the high element

chrisjbris committed:
Refactor to support the source modifiers on either or both operands.
Also extend the test.
Still struggling with 64-bit though as the legalizer is splitting
some 64-bit ops into v2i32.

chrisjbris commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4864,12 +4920,24 @@ SDValue AMDGPUTargetLowering::performSelectCombine(SDNode *N,
+
+    // Support source modifiers as integer.
+    if (VT == MVT::i32 || VT == MVT::v2i32 || VT == MVT::i64) {
+      if (SDValue SrcMod = getBitwiseToSrcModifierOp(True, DCI)) {
```
I've refactored and either or both operands can be folded now.

chrisjbris reviewed (commented):
ty - However I'm still struggling with i64. The DAG produces the correct operations but then ISEL lowers to bitwise or/xor/and after i64 has been moved to v2i32.

chrisjbris committed:
Fix Typo.

chrisjbris committed:
Respond to reviewer - Add i16 tests, simplify obtaining type

chrisjbris committed:
Inline bitcast node creation.

chrisjbris committed:
Add functional implementation for i64

While this is functional it can be refactored and simplified, working on
this now.

chrisjbris commented:
So the i64 fneg case is failing. However, i64 fabs and fneg+fabs are succeeding. Quick analysis shows a combine somewhere adding an extra fneg, creating a double negative, which cancels out. I'll try to identify the offending combine now.

chrisjbris committed:
Fix formatting

chrisjbris commented:
So what's happening here is, after the source modifier folding for the select has occurred, the combiner later produces:
```
Combining: t17: i64 = xor # D:1 t14, Constant:i64<-9223372036854775808>
Creating new node: t51: v2i32 = bitcast # D:1 t14
Creating new node: t52: i32 = extract_vector_elt # D:1 t51, Constant:i32<0>
Creating new node: t53: i32 = extract_vector_elt # D:1 t51, Constant:i32<1>
Creating constant: t54: i32 = Constant<-2147483648>
Creating new node: t55: i32 = xor # D:1 t53, Constant:i32<-2147483648>
Creating new node: t56: v2i32 = BUILD_VECTOR # D:1 t52, t55
Creating new node: t57: i64 = bitcast # D:1 t56
```
Presumably a realisation the only the high value of the xor operands are significant.
Later this i32 xor node is passed into DAGCombiner::foldBitcastedFPLogic() which creates:
```
Creating new node: t58: f32 = bitcast # D:1 t53
Creating new node: t59: f32 = fneg # D:1 t58
```
Resulting in fneg being applied twice. Hence, the vgpr in the fneg i64 test cases are not negated.

chrisjbris commented:
I'm currently attempting move a simplified version of the selectcombine implemented into the TI selectcombine, to take advantage of the combiner splitting the i64 op for us. This will require careful use of  shouldFoldSelectWithIdentityConstant() from TargetLowering.h.

chrisjbris committed:
[DAGCombine] Move the AMDGPU combine to Target Indepenent DAGCombine

- Allows removal of i64 specific code - the TI combine splits to i32
  ops.

- Update quite a few AMDGPU tests, these all appear to be improvements
  in codegen. Need to double-check.

chrisjbris committed:
Remove dead code that was moved to the TI DAGCombiner

chrisjbris committed:
Canonicalise TI select operand variable names and update tests

chrisjbris commented:
I've moved a simplified combine to the TI DAGCombiner::visitSelect(). For AMDGPU this appears to have improved codegen across a number of tests, but probably worth inspecting by more experience AMDGPU programmers. Thanks. but let's see what CI thinks, as locally I only test against X86 and AMDGPU.

chrisjbris committed:
Fix missed clang-format

chrisjbris committed:
Suppress overzealous clang-format

chrisjbris committed:
Suppress overzealous clang-format

chrisjbris committed:
Remove unnecessary lambda and refactor foldSelectOfSourceMods() to fit TI DAGCombiner style.

chrisjbris committed:
[NFC] Minor corrections to whitespace and test name

JanekvO commented on llvm/test/CodeGen/AMDGPU/fptrunc.ll:
```diff
@@ -111,34 +111,36 @@ define amdgpu_kernel void @fptrunc_f64_to_f16(ptr addrspace(1) %out, double %in)
 ; SI-NEXT:    s_bfe_u32 s0, s7, 0xb0014
```
I could be wrong but just looking at diff chunk changes in this test file, this test seem to regress (i.e., more lines added than removed)

JanekvO commented on llvm/test/CodeGen/AMDGPU/fneg-combines.new.ll:
```diff
@@ -4358,9 +4358,10 @@ define <2 x half> @v_fneg_select_infloop_regression_v2f16(<2 x half> %arg, i1 %a
-; SI-NEXT:    v_cmp_eq_u32_e32 vcc, 1, v1
-; SI-NEXT:    v_cndmask_b32_e64 v0, v0, 0, vcc
 ; SI-NEXT:    v_xor_b32_e32 v0, 0x80008000, v0
+; SI-NEXT:    v_mov_b32_e32 v2, 0x80008000
```
Why does this explicitly add the constant?

chrisjbris committed:
Add tighter constraints to apply combine and update tests.

chrisjbris committed:
Further constrain shouldFoldSelectWithIdentityConstant(), preventing regressions

chrisjbris committed:
Fix broken insert-delay-alu-bug.ll test

chrisjbris committed:
Replace target-specific function names with target-independent names
(Remove references to source modifiers).

chrisjbris commented:
Although moving to target independent combine enabled a simpler solution and resulted in improvements in other tests, I'm not sure this was the best way of working around the problem with the ordering of the combines. I'm considering other options.

chrisjbris commented:
Thanks to all reviewers but I think I need to deprecate this PR in favour of the simpler https://github.com/llvm/llvm-project/pull/149110.
